### PR TITLE
bios.rs: add `adopt-and-update`

### PR DESCRIPTION
**Pull request by HuijingHei**
Fixes https://github.com/coreos/bootupd/issues/717

### CODE REVIEW

**Comment by HuijingHei** [comment]
Hold on, need to deduplicate the code first.

File: src/bios.rs [hunk 1]
```diff
@@ -122,16 +122,24 @@ impl Component for Bios {
...
-        sysroot: &openat::Dir,
-        update: &ContentMetadata,
-    ) -> Result<InstalledContent> {
-        todo!();
+    fn adopt_update(&self, _: &openat::Dir, update: &ContentMetadata) -> Result<InstalledContent> {
+        let meta = if let Some(meta) = self.query_adopt()? {
+            meta
+        } else {
+            anyhow::bail!("Failed to find adoptable system");
+        };
```
Comment (cgwalters) [inline src/bios.rs:133]: Totally fine as is, we have a lot of this in the code but in newer Rust one can do instead:

`let Some(meta) = self.query_adopt()? else { anyhow::bail!("Failed to find adoptable system") };`
which is shorter. See e.g. https://github.com/containers/bootc/pull/650/files

File: src/component.rs [hunk 2]
```diff
@@ -144,6 +144,38 @@ pub(crate) fn get_component_update(
     }
 }
 
+#[context("query adopt impl")]
```
Comment (cgwalters) [inline src/component.rs:147]: The intention for `context` is to provide something friendly for an external user. Maybe `Querying adoptable state`?

File: src/component.rs [hunk 3]
```diff
@@ -144,6 +144,38 @@ pub(crate) fn get_component_update(
...
+pub(crate) fn query_adopt_impl() -> Result<Option<Adoptable>> {
+    // This would be extended with support for other operating systems later
+    if let Some(coreos_aleph) = crate::coreos::get_aleph_version(Path::new("/"))? {
+        let meta = ContentMetadata {
+            timestamp: coreos_aleph.ts,
+            version: coreos_aleph.aleph.version,
+        };
+        log::trace!("BIOS adoptable: {:?}", &meta);
```
Comment (cgwalters) [inline src/component.rs:155]: I think we can just remove "BIOS" here

**Review by cgwalters — APPROVED** [review]
Overall looks sane to me, just a few nits.
Thanks for working on this!

File: src/bios.rs (hunk 1 above)
Comment (HuijingHei) [reply src/bios.rs]: Will update this in later patch, thanks!

**Review by cgwalters — APPROVED** [review]

**Pull request merged**